arctic/async/async_arctic.py

```diff
@@ -0,0 +1,187 @@
+            self.requests_by_id = dict()
+
+    def __reduce__(self):
+        return "ASYNC_ARCTIC"
+
+    def _get_modifiers(self, library_name, symbol=None):
+        return self.requests_per_library[library_name][symbol][AsyncRequestType.MODIFIER]
+
+    def _get_accessors(self, library_name, symbol=None):
+        return self.requests_per_library[library_name][symbol][AsyncRequestType.ACCESSOR]
+
+    @staticmethod
+    def _verify_request(store, is_modifier, **kwargs):
+        library_name = None if store is None else store._arctic_lib.get_name()
+        symbol = kwargs.get('symbol')
+        kind = AsyncRequestType.MODIFIER if is_modifier else AsyncRequestType.ACCESSOR
+        callback = kwargs.get('async_callback')
+        mongo_retry = bool(kwargs.get('mongo_retry'))
+        return library_name, symbol, kind, callback, mongo_retry
+
+    def _add_request(self, request):
+        self.requests_per_library[request.library][request.symbol][request.kind].append(request)
+        self.requests_by_id[request.id] = request
+
+    def _remove_request(self, request):
+        self.requests_per_library[request.library][request.symbol][request.kind].remove(request)
+        if request.id in self.requests_by_id:
+            del self.requests_by_id[request.id]
+
+    def submit_arctic_request(self, store, fun, is_modifier, *args, **kwargs):
+        lib_name, symbol, kind, callback, mongo_retry = AsyncArctic._verify_request(store, is_modifier, **kwargs)
+
+        for k in ('async_callback', 'mongo_retry'):
+            kwargs.pop(k, None)
+
+        with type(self)._POOL_LOCK:  # class level lock, since it is a Singleton
+            if lib_name:
+                if self._get_modifiers(lib_name, symbol):
+                    raise AsyncArcticException("Can't submit async task as one or more {} tasks "
+                                               "are already being processed".format(AsyncRequestType.MODIFIER))
+
+                if is_modifier and self._get_accessors(lib_name, symbol):
+                    raise AsyncArcticException("Can't submit async {} task as one or more {} tasks "
+                                               "are being processed".format(AsyncRequestType.ACCESSOR,
+                                                                            AsyncRequestType.MODIFIER))
+
+            # Create the request object
+            request = AsyncRequest(kind, lib_name, fun, *args, **kwargs)
+
+            # Update the state of tracked tasks
+            self._add_request(request)
+
+            # Submit the task
+            try:
+                new_id, new_future = self.submit_task(False, _arctic_task_exec, request)
+                request.id = new_id
+                request.future = new_future
+            except Exception:
+                # clean up the state
+                self._remove_request(request)
+                raise
```

Comment:
what exceptions do you expect here?  backpressure from the executor?  Should they be returned through the callback?

Reply:
this is scheduling related exceptions, not related to the task we are supposed to run, becuase all task exceptions are attached to the request object.

This means that the executor service failed to process/run the request, not that the request's body excecution failed.

we should be really raising this one